### PR TITLE
Transient Optimization problem formulation

### PR DESCRIPTION
**Pull request by kaarthiksundar**
This PR includes the following:

1. Parsing transient data file along with static file using the command ``parse_files(static_file, transient)``
2. Creates ``time_series_block`` with interpolation capability for multinetwork creation. 
3. ``build_transient_opf`` problem formulates the transient optimization problem - supports only pipes, junctions, compressors, receipts, deliveries and transfers. No other components are supported. Objective is a linear combination of both the compressor power and load shed. 
4. Revamped the unit conversions to support (memory efficient) multi-network unit conversions.
5. Added a ref extension for the transient optimization problem - this adds all the additional fields required to formulate the problem.
6. Variables for the transient problem are in transient_variable.jl 
7. Expressions for the transient problem are in transient_expression.jl
8. Constraints for transient problem are organized int constraint_template_transient.jl and constraint_transient.jl 
9. Added solution reporting for the full transient model.
10. Added unit test for the transient problem - one is a constant steady state test case and another is a time-periodic withdrawal test case.
11. Objective has been split into load shed, compressor power and combined cases.
12. Added more fields into unit conversions for cross-compatibility with results. 

Bug fixes:
1. Fixed bug in area computation - modified the tests accordingly. 
2. Fixed bug in sound speed computation and other parameters
3. Fixed bugs in matgas parser by merging parses-fixes branch by @pseudocubic

### CODE REVIEW

File: src/core/transient_variable.jl [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        for (i, junction) in ref(gm, nw, :junction)
+            JuMP.set_lower_bound(rho[i], junction["p_min"])
+            JuMP.set_upper_bound(rho[i], junction["p_max"])
+        end
+    end
+
+    report &&
+    _IM.sol_component_value(gm, nw, :junction, :density, ids(gm, nw, :junction), rho)
+    report &&
+    _IM.sol_component_value(gm, nw, :junction, :pressure, ids(gm, nw, :junction), rho)
```
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply src/core/transient_variable.jl:31]: No, non-dimensional pressure and density are equal when you use the ideal gas assumption. If not, they are related by a function.

File: test/data/transient/old_format/time_series.csv [hunk 2]
```diff
@@ -0,0 +1,101 @@
+2020-01-01T00:00:00
```
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply test/data/transient/old_format/time_series.csv:1]: We don't need this file - will remove.

File: src/io/matgas.jl [hunk 3]
```diff
@@ -328,25 +354,36 @@ function parse_m_string(data_string::String)
...
     if haskey(matlab_data, "mgc.junction")
         junctions = []
         for junction_row in matlab_data["mgc.junction"]
-            junction_data = InfrastructureModels.row_to_typed_dict(junction_row, _mg_junction_columns)
-            junction_data["index"] = InfrastructureModels.check_type(Int, junction_row[1])
+            junction_data = _IM.row_to_typed_dict(junction_row, get(_colnames, "mgc.junction", _mg_junction_columns))
+            junction_data["index"] = _IM.check_type(Int, junction_row[1])
+            junction_data["is_si_units"] = case["is_si_units"]
+            junction_data["is_english_units"] = case["is_english_units"]
+            junction_data["is_per_unit"] = case["is_per_unit"]
```
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply src/io/matgas.jl:386]: I agree, but I was not able to come up with any other way without introducing logical bugs.

File: src/core/constraint_template_transient.jl [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    fixed_density = ref(gm, nw, :slack_junctions, i)["p_nominal"]
+    constraint_slack_junction_density(gm, nw, i, fixed_density)
+end
+
+"slack junction mass balance"
+function constraint_slack_junction_mass_balance(
+    gm::AbstractGasModel,
+    i::Int,
+    nw::Int = gm.cnw,
+)
```
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply src/core/constraint_template_transient.jl:15]: I am just using JuliaFormatter.jl - it auto formats; have the vscode plugin setup.

File: src/core/transient_variable.jl [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    if report
+        _IM.sol_component_value(gm, nw, :pipe, :flux, ids(gm, nw, :pipe), phi)
+        sol_f =
+            Dict(i => phi[i] * ref(gm, nw, :pipe, i)["area"] for i in ids(gm, nw, :pipe))
+        _IM.sol_component_value(gm, nw, :pipe, :flow, ids(gm, nw, :pipe), sol_f)
+    end
+end
+
+"variables associated with compression ratio (transient)"
+function variable_c_ratio(
```
[the remark this answers is not in the document]
Comment (rb004f) [reply src/core/transient_variable.jl:102]: In general, I'd like to see the transient model reusing as much as possible from the steady-state model, so if this can directly use the compression ratio variable in ``variable.jl`` lets do that.  In the long run, esp. as we get a steady-state more in the density/flux space we'll have alot of overlap and I'd like to leverage that is much as possible to keep the code small and agile.

File: src/core/constraint_template_transient.jl [hunk 6]
```diff
@@ -0,0 +1,81 @@
+"fixing slack node density value"
+function constraint_slack_junction_density(gm::AbstractGasModel, i::Int, nw::Int = gm.cnw)
+    if !haskey(gm.con[:nw][nw], :slack_junction_density)
+        con(gm, nw)[:slack_junction_density] = Dict{Int,JuMP.ConstraintRef}()
+    end
+    fixed_density = ref(gm, nw, :slack_junctions, i)["p_nominal"]
+    constraint_slack_junction_density(gm, nw, i, fixed_density)
+end
+
```
[the remark this answers is not in the document]
Comment (rb004f) [reply src/core/constraint_template_transient.jl:9]: I recommend using the ``add_constraint function`` when you create a constraints.  I created this function to automatically store constraints which are created in the constraint store, i.e. just do ``_add_constraint!(gm, nw, :slack_junction_density, i, JuMP.@constraint(foo))`` when you create the constraint and you can safely delete lines 3-5.

File: src/core/base.jl [hunk 7]
```diff
@@ -176,6 +177,7 @@ function _ref_add_core!(nw_refs::Dict{Int,<:Any}; base_length=5000.0, base_press
         for (idx, compressor) in ref[:compressor]
             i = compressor["fr_junction"]
             j = compressor["to_junction"]
+            compressor["area"] = pi * compressor["diameter"]  * compressor["diameter"] / 4.0
```
Comment (rb004f) [inline src/core/base.jl:180]: I like adding area calculations.  We should now use this directly in other parts of the code (i.e. the resistance calculations) rather than recompute

File: src/core/constraint_template_transient.jl [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+"slack junction mass balance"
+function constraint_slack_junction_mass_balance(
+    gm::AbstractGasModel,
+    i::Int,
+    nw::Int = gm.cnw,
+)
+    if !haskey(gm.con[:nw][nw], :slack_junction_mass_balance)
+        con(gm, nw)[:slack_junction_mass_balance] = Dict{Int,JuMP.ConstraintRef}()
+    end
+    net_injection = var(gm, nw, :net_nodal_injection)[i]
```
Comment (rb004f) [inline src/core/constraint_template_transient.jl:19]: see comment above about ``add_constraint`` function

File: src/core/constraint_template_transient.jl [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    to_junction = pipe["to_junction"]
+    resistance =
+        pipe["friction_factor"] * gm.ref[:base_length] * pipe["length"] / pipe["diameter"]
+    constraint_pipe_physics_ideal(gm, nw, i, fr_junction, to_junction, resistance)
+end
+
+"compressor physics"
+function constraint_compressor_physics(gm::AbstractGasModel, i::Int, nw::Int = gm.cnw)
+    if !haskey(gm.con[:nw][nw], :compressor_physics_boost)
+        con(gm, nw)[:compressor_physics_boost] = Dict{Int,JuMP.ConstraintRef}()
```
Comment (rb004f) [inline src/core/constraint_template_transient.jl:62]: This looks like the upper and lower bound constraints on compression ratios.  Can we use the steady declaration of this constraint directly?

File: src/core/constraint_template_transient.jl [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        con(gm, nw)[:compressor_physics_flow] = Dict{Int,JuMP.ConstraintRef}()
+    end
+    compressor = ref(gm, nw, :compressor, i)
+    fr_junction = compressor["fr_junction"]
+    to_junction = compressor["to_junction"]
+    constraint_compressor_physics(gm, nw, i, fr_junction, to_junction)
+end
+
+"compressor power"
+function constraint_compressor_power(gm::AbstractGasModel, i::Int, nw::Int = gm.cnw)
```
Comment (rb004f) [inline src/core/constraint_template_transient.jl:74]: This looks the same as the compresor power constraint for steady state.  Can we use directly and delete this one?

File: src/core/constraint_transient.jl [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    s = var(gm, nw, :transfer_injection)[transfer_id]
+    d = var(gm, nw, :transfer_withdrawal)[transfer_id]
+    t = var(gm, nw, :transfer_effective)[transfer_id]
+
+    con(gm, nw, :effective_transfer_withdrawal)[transfer_id] =
+        JuMP.@constraint(gm.model, t == d - s)
+end
+
+"compressor physics"
+function constraint_compressor_physics(
```
Comment (rb004f) [inline src/core/constraint_transient.jl:74]: This constraint is implemented the same for steady-state?

File: src/core/constraint_transient.jl [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    alpha = var(gm, nw, :compressor_ratio, compressor_id)
+    f = var(gm, nw, :compressor_flow, compressor_id)
+    con(gm, nw, :compressor_physics_boost)[compressor_id] =
+        JuMP.@constraint(gm.model, p_to == alpha * p_fr)
+    con(gm, nw, :compressor_physics_flow)[compressor_id] =
+        JuMP.@constraint(gm.model, f * (p_fr - p_to) <= 0)
+end
+
+"compressor power"
+function constraint_compressor_power(
```
Comment (rb004f) [inline src/core/constraint_transient.jl:92]: constraint is the same in steady-state?

File: src/core/objective.jl [hunk 13]
```diff
@@ -1,71 +1,217 @@
...
+            sum(-transfer_prices[n][i] * ft[n][i] for i in transfer_set[n]) +
+            economic_weighting * sum(prod_prices[n][i] * fg[n][i] for i in prod_set[n]) +
+            (1.0 - economic_weighting) *
+            sum(f[n][i] * (r[n][i]^m - 1) for (i, compressor) in ref(gm, n, :compressor))
+            for n in nws
+        )
+    )
+end
+
+function objective_transient_economic(gm::AbstractGasModel, time_points)
```
[the remark this answers is not in the document]
Comment (rb004f) [reply src/core/objective.jl:132]: Is the transient economic costs the same as the steady state model? The steady-state was already looping through all networks, so maybe we don't need a separate transient one?

File: src/core/objective.jl [hunk 14]
```diff
@@ -1,71 +1,217 @@
...
+        Min,
+        econ_weight *
+        sum(load_shed_expressions[i] for i = 1:length(load_shed_expressions)) +
+        (1 - econ_weight) *
+        sum(compressor_power_expressions[i] for i = 1:length(compressor_power_expressions))
+    )
+    end
+end
+
+function objective_transient_load_shed(gm::AbstractGasModel, time_points)
```
[the remark this answers is not in the document]
Comment (rb004f) [reply src/core/objective.jl:181]: see same comment on economic costs

File: test/data/matgas/A2.m [hunk 15]
```diff
@@ -15,7 +15,6 @@
 mgc.is_per_unit                  = 0;
```
Comment (rb004f) [inline test/data/matgas/A2.m:15]: I modified this file to improve formatting and numerics in the examples pull request.  I think you can safely discard these changes in lue of what was merged into master there

File: test/data/matgas/case-6-no-power-limits.m [hunk 16]
```diff
@@ -55,5 +55,5 @@
 %% receipt data
 % id junction_id injection_min injection_max injection_nominal is_dispatchable status offer_price name company_name daily_scheduled_flow design_capacity operating_capacity is_firm edi_id
 mgc.receipt = [
-1    1    0    1000    500    1     1  -1000
```
Comment (rb004f) [inline test/data/matgas/case-6-no-power-limits.m:58]: why this change?  To match @azlotnik model?

File: test/data/matgas/gaslib-135.m [hunk 17]
```diff
@@ -13,6 +13,7 @@
 mgc.base_pressure = 8101325;  % Pa
 mgc.base_length = 5000.0000;  % m
 mgc.base_flow = 1100.0;
```
Comment (rb004f) [inline test/data/matgas/gaslib-135.m:15]: I modified this file a lot to improve its numerics when examples was merged.  I think you can safely discard these changes

File: test/ogf.jl [hunk 18]
```diff
@@ -1,23 +1,23 @@
 @testset "test ogf" begin
     @testset "test nlp ogf" begin
-        @testset "gaslib 40 nlp ogf" begin
+        @testset "case 6 ogf" begin
```
Comment (rb004f) [inline test/ogf.jl:3]: why did the objective value change on these problems?

File: .travis.yml [hunk 19]
```diff
@@ -4,15 +4,15 @@ os:
...
+  - 1.4
   - nightly
 codecov: true
 jobs:
   allow_failures:
     - julia: nightly
   include:
     - stage: "Documentation"
-      julia: 1.3
+      julia: 1.4
```
Comment (rb004f) [inline .travis.yml:15]: I added an examples stage to this script, so a rebase with master will be necessary

**Comment by rb004f** [comment]
@kaarthiksundar @pseudocubic Everything looks great to me.  I also only had very minor comments.  The only remotely large comment is to suggest we reuse the steady-state declaration of constraints, variables, and objectives where they are already identical with the transient model.  Those declarations already either loop through all the networks to create variables and objectives or allow you to pass in the network index as a parameter.  We can't do this for constraints in density or flux space, but we can do it for things like compression ratio variables or compressor power constraints.  Otherwise, I think this looks good to merge

File: CHANGELOG.md [hunk 20]
```diff
@@ -8,6 +8,10 @@
 ## staged
 
 - Update to InfrastructureModels solution building (breaking)
+- Add support for Memento v0.13, v1.0
+- Introduce transient optimal gas flow formulation (transient_ogf)
+- Support for transient optimization problem restricted to component types: junction, pipe, compressor, receipt, delivery, and transfers; objective support: load shed, compressor power, and linear combination of both.
+- Add multi-network and transient_ogf unit tests
```
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply CHANGELOG.md:14]: Done

File: src/core/variable.jl [hunk 21]
```diff
@@ -194,10 +194,10 @@ function variable_load_mass_flow(gm::AbstractGasModel, nw::Int=gm.cnw; bounded::
     end
 
     if report
-        _IM.sol_component_value(gm, nw, :delivery, :fl, ids(gm, nw, :dispatchable_delivery), fl)
+        _IM.sol_component_value(gm, nw, :delivery, :fd, ids(gm, nw, :dispatchable_delivery), fl)
         if haskey(gm.data, "standard_density")
             sol_ql = Dict(i => fl[i] / gm.data["standard_density"] for i in ids(gm, nw, :dispatchable_delivery))
-            _IM.sol_component_value(gm, nw, :delivery, :ql, ids(gm, nw, :dispatchable_delivery), sol_ql)
+            _IM.sol_component_value(gm, nw, :delivery, :qd, ids(gm, nw, :dispatchable_delivery), sol_ql)
```
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply src/core/variable.jl:200]: I think here the convention was generation and demand (:fg and :fd)

File: src/core/transient_variable.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply src/core/transient_variable.jl:102]: Agree, but as @pseudocubic said, there is no compressor ratio variable in variable.jl. There is only compressor ratio squared.

[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply src/core/transient_variable.jl:102]: Refactored the function name in variable.jl to ``variable_compressor_ratio_sqr`` and ``variable_compresor_ratio_sqr_value``

File: src/io/transient.jl [hunk 22]
```diff
@@ -13,8 +13,349 @@ function parse_transient(io::IO)::Array{Dict{String,Any},1}
...
+                "is_bidirectional" => pipe["is_bidirectional"],
+                "is_si_units" => data["is_si_units"],
+                "is_english_units" => data["is_english_units"],
+                "is_per_unit" => data["is_english_units"],
+            )
+        end
+    end
+end
+
+function create_time_series_block(
```
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply src/io/transient.jl:241]: Yes it is private. Added a ``_`` and created a docstring.

File: src/prob/transient_ogf.jl [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+    var(gm, end_t)[:density] = var(gm, time_points[start_t], :density)
+    var(gm, end_t)[:compressor_flow] = var(gm, time_points[start_t], :compressor_flow)
+    var(gm, end_t)[:pipe_flux] = var(gm, time_points[start_t], :pipe_flux)
+    var(gm, end_t)[:compressor_ratio] = var(gm, time_points[start_t], :compressor_ratio)
+    var(gm, end_t)[:injection] = var(gm, time_points[start_t], :injection)
+    var(gm, end_t)[:withdrawal] = var(gm, time_points[start_t], :withdrawal)
+    var(gm, end_t)[:transfer_effective] = var(gm, time_points[start_t], :transfer_effective)
+    var(gm, end_t)[:transfer_injection] = var(gm, time_points[start_t], :transfer_injection)
+    var(gm, end_t)[:transfer_withdrawal] =
+        var(gm, time_points[start_t], :transfer_withdrawal)
```
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply src/prob/transient_ogf.jl:41]: I spoke to @ccoffrin about this - this is pretty much his suggestion as well.

File: src/core/base.jl (hunk 7 above)
Comment (kaarthiksundar) [reply src/core/base.jl:180]: @rb004f converted this.

File: src/core/constraint_template_transient.jl (hunk 10 above)
Comment (kaarthiksundar) [reply src/core/constraint_template_transient.jl:74]: @rb004f Yes, but there is something different with the way we do the non-dimensionalization. I was not able to reproduce results. So I decided to write my own. Will get to this issue at the end of this sprint.

File: src/core/constraint_transient.jl (hunk 11 above)
Comment (kaarthiksundar) [reply src/core/constraint_transient.jl:74]: @rb004f yes and no. The constraints that are implemented live in the ratio_sqr and p_sqr space.

File: test/data/matgas/case-6-no-power-limits.m (hunk 16 above)
Comment (kaarthiksundar) [reply test/data/matgas/case-6-no-power-limits.m:58]: Yes.

File: test/ogf.jl (hunk 18 above)
Comment (kaarthiksundar) [reply test/ogf.jl:3]: I fixed some bugs in the per unit conversions and changed certain values in the data files to match old results.

File: src/core/constraint_template_transient.jl (hunk 10 above)
Comment (rb004f) [reply src/core/constraint_template_transient.jl:74]: Ok, note as an issue and we'll come back to it.

File: src/core/constraint_transient.jl (hunk 11 above)
Comment (rb004f) [reply src/core/constraint_transient.jl:74]: Understood.  We should probably adopt a naming convention to make that clear now that we are using these variables in both the linear and square spaces

File: test/data/matgas/case-6-no-power-limits.m (hunk 16 above)
Comment (rb004f) [reply test/data/matgas/case-6-no-power-limits.m:58]: Ok, I had matched this to a model/results he was using in his matlab code.  If that has changed, then I am fine with this modification.

**Comment by rb004f** [comment]
@pseudocubic @kaarthiksundar Anything else we need to do a merge?  I think once there is documentation for the transient model, I think it would make sense to tag a new release as well.

**Comment by kaarthiksundar** [comment]
> @pseudocubic @kaarthiksundar Anything else we need to do a merge? I think once there is documentation for the transient model, I think it would make sense to tag a new release as well.

I have to push the changes. Will do in an hour.

File: src/core/constraint_template_transient.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (kaarthiksundar) [reply src/core/constraint_template_transient.jl:9]: Done, this functionality is very nice - thanks @rb004f

File: src/core/constraint_transient.jl (hunk 12 above)
Comment (kaarthiksundar) [reply src/core/constraint_transient.jl:92]: Added issue

**Comment by kaarthiksundar** [comment]
@pseudocubic @rb004f Addressed all the reviews. Feel free to merge once travis passes. 

Thanks

**Comment by rb004f** [comment]
@kaarthiksundar looks like you are still pushing a few changes... let me know when you are done and we can merge.

**Comment by kaarthiksundar** [comment]
> @kaarthiksundar looks like you are still pushing a few changes... let me know when you are done and we can merge.

@rb004f Some tests failed - pushed some commits to fix them. But now having issues with travis builds.

**Comment by kaarthiksundar** [comment]
@pseudocubic @rb004f All tests pass, you can merge.
